wordpress-mobile/release-toolkit
# Add toolkit-autoupdate action

## Why?

Now that the release toolkit uses semver, I started to [introduce an auto-update mechanism for it in the woocommerce repositories](https://github.com/woocommerce/woocommerce-ios/pull/4288/files#diff-dff4b99d4e651ab9d7597876ec8dbe92aa934e42c0fb55f4aadd6c106fc96688R20-R32).

As the check logic implemented in WC's Fastfile still took a couple of lines, @oguzkocer [rightfully pointed out](https://github.com/woocommerce/woocommerce-android/pull/4089#discussion_r639923075) that we would benefit from making this into an action, for other apps to be able to benefit it. That's what this PR is all about.

## How

Inspired by [fastlane's own action to auto-update fastlane itself](https://github.com/fastlane/fastlane/blob/master/fastlane/lib/fastlane/actions/update_fastlane.rb), it checks for the locally-installed version and if there is a more recent update was available.

If there is one – be it a breaking change or not – it warns you about the latest version available, then propose to run `bundle update` for you. If you say yes, it will run `bundle update` then stop the lane – both in order to let you commit the changes and to force you to restart your fastlane invocation after the update so that fastlane picks up the updated plugin.

## To Test

This is gonna be tricky to test, because this is gonna be the first version of the toolkit that is gonna introduce this action, so it's not like we could point to an older version and test that it detects a new one…

The best thing I was able to do is to temporarily tweak the action to check for another gem instead:

 * Temporarily modify the `TOOLKIT_SPEC_NAME` constant in that branch, to be another random gem we use in our app repos (e.g. `fastlane`, `nokogiri`, `fastlane-plugin-appcenter`, …)
 * Commit and push the changes on that branch
 * In an app repo, edit the `Pluginfile` to point the toolkit to `git: …, branch: 'action/autoupdate'`, and run `bundle install`
 * Add a call to `check_for_toolkit_updates unless is_ci` in the `before_all` block of your `Fastfile`
 * Ensure that the dummy gem we chose on step 1 for this test is not yet on the latest version in that app repo. If necessary, update your `Gemfile` to temporarily force installing an older version of that dummy gem instead.
 * Run an arbitrary (and ideally side-effects-free) lane from the app repo (e.g. `bundle exec fastlane get_pullrequests_list` or `bundle exec fastlane download_promo_strings`), and check that the `check_for_toolkit_updates` action gets called.
 * Check that it warns you about the dummy gem being out of date. Refuse to update when prompted, and see it continue with the lane
 * Run the same lane again, and this time accept to run bundle update when prompted. See it update your `Gemfile.lock` and stop the lane, telling you to commit and restart fastlane.
 * Run the same lane one more time, and ensure that this time there is no new update found (providing the version constraint for the dummy gem you used in `Gemfile` wasn't too restrictive to get that latest version)
 * Revert the commit you made on step 1-2 (the temporary change to `TOOLKIT_SPEC_NAME`)
